module/apms3session/session.go

```diff
@@ -0,0 +1,153 @@
+)
+
+func init() {
+	stacktrace.RegisterLibraryPackage(
+		"github.com/aws/aws-sdk-go/aws/request",
+		"github.com/aws/aws-sdk-go/aws/session",
+	)
+}
+
+type handlers struct {
+	tracer *apm.Tracer
+}
+
+// Option sets options for tracing s3 requests.
+type Option func(*handlers)
+
+// WithTracer returns an Option which sets t as the tracer to use for tracing
+// s3 requests.
+func WithTracer(t *apm.Tracer) Option {
+	if t == nil {
+		panic("t == nil")
+	}
+	return func(h *handlers) {
+		h.tracer = t
+	}
+}
+
+// WrapSession wraps the provided s3 session with handlers that hook into the
+// aws sdk's request lifecycle.
+func WrapSession(s *session.Session, opts ...Option) *session.Session {
+	h := &handlers{
+		tracer: apm.DefaultTracer,
+	}
+	for _, o := range opts {
+		o(h)
+	}
+
+	s.Handlers.Send.PushFrontNamed(request.NamedHandler{
+		Name: "go.elastic.co/apm/module/apms3session/send",
+		Fn:   h.send,
+	})
+	s.Handlers.Complete.PushBackNamed(request.NamedHandler{
+		Name: "go.elastic.co/apm/module/apms3session/complete",
+		Fn:   h.complete,
+	})
+
+	return s
+}
+
+const (
+	spanType    = "storage"
+	spanSubtype = "s3"
+)
+
+func (h *handlers) send(req *request.Request) {
+	if req.RetryCount > 0 {
+		return
+	}
+
+	var (
+		bucketName string
+
+		ctx    = req.Context()
+		region = *req.Config.Region
+	)
+
+	tx := apm.TransactionFromContext(ctx)
+	if tx == nil {
+		return
+	}
+
+	// TODO: All Params structs are required to have a `Bucket` attr. If
+	// there's a better way to get this, I would be overjoyed to avoid
+	// using runtime reflection.
+	params := reflect.ValueOf(req.Params).Elem()
+	if n, ok := params.FieldByName("Bucket").Interface().(*string); ok {
+		bucketName = *n
+	} else {
+		// TODO: How do we want to handle this?
+		fmt.Printf("couldn't reflect Bucket into string %+v\n", req.Params)
+	}
+
+	spanName := spanSubtype + " " + req.Operation.Name + " " + bucketName
+	span := tx.StartSpan(spanName, spanType, apm.SpanFromContext(ctx))
+	if !span.Dropped() {
+		ctx = apm.ContextWithSpan(ctx, span)
+		req.HTTPRequest = apmhttp.RequestWithContext(ctx, req.HTTPRequest)
+		span.Context.SetHTTPRequest(req.HTTPRequest)
+	} else {
+		span.End()
+		span = nil
+	}
+
+	span.Subtype = spanSubtype
+	span.Action = req.Operation.Name
+
+	destinationAddress := req.ClientInfo.Endpoint
+	span.Context.SetDestinationAddress(destinationAddress, 0)
+	span.Context.SetDestinationService(apm.DestinationServiceSpanContext{
+		Name:     spanSubtype,
+		Resource: bucketName,
+		Type:     spanType,
+	})
+	span.Context.SetDestinationCloud(apm.DestinationCloudSpanContext{
+		Region: region,
+	})
+
+	req.SetContext(ctx)
+}
+
+func (h *handlers) complete(req *request.Request) {
+	ctx := req.Context()
+	span := apm.SpanFromContext(ctx)
+	if span.Dropped() {
+		return
+	}
+	defer span.End()
+
+	span.Context.SetHTTPStatusCode(req.HTTPResponse.StatusCode)
+
+	if err := req.Error; err != nil {
+		apm.CaptureError(ctx, err).Send()
+	}
```

Comment:
Are these definitely exceptional errors, or could they be expected errors used for control flow? e.g. create bucket -> already exists -> not really an error

Reply:
I think it could be something as simple as a bucket already existing, but it could also be that the user for the session doesn't have perms to perform operation X, which might be good to surface? if access changed for some reason, it would definitely pinpoint the failure reason pretty quick.

Reply:
OK. I'm alright adding them by default, maybe later we'll add an option to suppress some errors. Can you please check with other agent devs to see what if anything they're doing?

Reply:
According to the node.js agent, they're reporting errors where the SDK ran into things like an authentication issue, a transport issue, or returned something error like, but not reporting errors that are related to the queue operation itself (bucket already exists, too many message attributes). There's no official position and they're the only people I've heard back from, but we can at least be consistent with them.

Reply:
sounds good to me